contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -380,60 +452,73 @@ contract LidoOracle is ILidoOracle, AragonApp {
+            if (i != maxi && gatheredReportKinds[i].count == gatheredReportKinds[maxi].count) {
+                return (false, Report({beaconBalance: 0, beaconValidators: 0}));
             }
         }
-
-        assert(i == data.length);
-
-        // find mode value of this array
-        (bool isUnimodal, uint256 mode) = Algorithm.mode(data);
-        if (!isUnimodal)
-            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
-
-        // unpack Report struct from uint256
-        modeReport = uint256ToReport(mode);
-
-        return (true, modeReport);
+        return (gatheredReportKinds[maxi].count >= quorum, gatheredReportKinds[maxi].report);
     }
 
     /**
      * @dev Pushes the current data point if quorum is reached
      */
-    function _tryPush(uint256 _epochId) internal {
-        (bool isQuorum, Report memory modeReport) = _getQuorumReport(_epochId);
-        if (!isQuorum)
-            return;
+    function _tryPush() internal {
+        (bool isQuorum, Report memory modeReport) = _getQuorumReport();
+        if (!isQuorum) return;
 
         // data for this frame is collected, now this frame is completed, so
         // minReportableEpochId should be changed to first epoch from next frame
         BeaconSpec memory beaconSpec = _getBeaconSpec();
-        MIN_REPORTABLE_EPOCH_ID_POSITION.setStorageUint256(
-            _epochId
+        uint256 epochId = REPORTABLE_EPOCH_ID_POSITION.getStorageUint256();
+        emit Completed(epochId, modeReport.beaconBalance, modeReport.beaconValidators);
+        _clearVotingAndAdvanceTo(
+            epochId
             .div(beaconSpec.epochsPerFrame)
             .add(1)
-            .mul(beaconSpec.epochsPerFrame)
-        );
-
-        emit Completed(_epochId, modeReport.beaconBalance, modeReport.beaconValidators);
+            .mul(beaconSpec.epochsPerFrame));
 
+        // report to the lido and collect stats
         ILido lido = getLido();
-        if (address(0) != address(lido))
-            lido.pushBeacon(modeReport.beaconValidators, modeReport.beaconBalance);
+        PRE_COMPLETED_TOTAL_POOLED_ETHER_POSITION.setStorageUint256(ISTETH(lido).totalSupply());
+        lido.pushBeacon(modeReport.beaconValidators, modeReport.beaconBalance);
+        POST_COMPLETED_TOTAL_POOLED_ETHER_POSITION.setStorageUint256(ISTETH(lido).totalSupply());
+        PREV_COMPLETED_EPOCH_ID_POSITION.setStorageUint256(LAST_COMPLETED_EPOCH_ID_POSITION.getStorageUint256());
+        LAST_COMPLETED_EPOCH_ID_POSITION.setStorageUint256(epochId);
+        reportSanityChecks();  // rollback on boundaries violation (logical consistency)
+
+        // call the quorum delegate
+        IQuorumCallback quorumCallbackAddr = IQuorumCallback(QUORUM_CALLBACK_POSITION.getStorageUint256());
+        if (address(quorumCallbackAddr) != address(0)) {
+            (uint256 lastTotalPooledEther,
+             uint256 prevTotalPooledEther,
+             uint256 timeElapsed) = getLastCompletedReports();
```

Comment:
Looks we are abusing public `getLastCompletedReports`: calling it here and inside `reportSanityChecks` just above.  Need to optimise it.

Reply:
Fix in https://github.com/lidofinance/lido-dao/pull/306...